src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4620 @@
+        )
+        self.update_block = ProPainterBasicUpdateBlock(config, hidden_dim=self.hidden_dim)
+
+    def initialize_flow(self, image):
+        """Flow is represented as difference between two coordinate grids flow = coords1 - coords0"""
+        N, _, height, width = image.shape
+        coords0 = coords_grid(N, height // 8, width // 8).to(image.device)
+        coords1 = coords_grid(N, height // 8, width // 8).to(image.device)
+
+        # optical flow computed as difference: flow = coords1 - coords0
+        return coords0, coords1
+
+    def upsample_flow(self, flow, mask):
+        """Upsample flow field [height/8, width/8, 2] -> [height, width, 2] using convex combination"""
+        N, _, height, width = flow.shape
+        mask = mask.view(N, 1, 9, 8, 8, height, width)
+        mask = torch.softmax(mask, dim=2)
+
+        up_flow = F.unfold(8 * flow, [3, 3], padding=self.config.padding)
+        up_flow = up_flow.view(N, 2, 9, 1, 1, height, width)
+
+        up_flow = torch.sum(mask * up_flow, dim=2)
+        up_flow = up_flow.permute(0, 1, 4, 2, 5, 3)
+        return up_flow.reshape(N, 2, 8 * height, 8 * width)
+
+    def _forward(self, image1, image2, iters=12, flow_init=None):
+        """Estimate optical flow between pair of frames"""
+
+        image1 = image1.contiguous()
+        image2 = image2.contiguous()
+
+        feature_map_1, feature_map_2 = self.feature_network([image1, image2])
+
+        feature_map_1 = feature_map_1.float()
+        feature_map_2 = feature_map_2.float()
+
+        correlation_fn = ProPainterCorrBlock(
+            self.config, feature_map_1, feature_map_2, radius=self.config.correlation_radius
+        )
+
+        context_network_out = self.context_network(image1)
+        network, input = torch.split(context_network_out, [self.hidden_dim, self.context_dim], dim=1)
+        network = torch.tanh(network)
+        input = torch.relu(input)
+
+        coords0, coords1 = self.initialize_flow(image1)
+
+        if flow_init is not None:
+            coords1 = coords1 + flow_init
+
+        for _ in range(iters):
+            coords1 = coords1.detach()
+            correlation = correlation_fn(coords1)  # index correlation volume
+
+            optical_flow = coords1 - coords0
+            network, up_mask, delta_flow = self.update_block(network, input, correlation, optical_flow)
+
+            coords1 = coords1 + delta_flow
+
+            if up_mask is None:
+                new_size = (
+                    8 * (coords1 - coords0).shape[2],
+                    8 * (coords1 - coords0).shape[3],
+                )
+                flow_up = 8 * F.interpolate(
+                    (coords1 - coords0),
+                    size=new_size,
+                    mode="bilinear",
+                    align_corners=True,
+                )
+            else:
+                flow_up = self.upsample_flow(coords1 - coords0, up_mask)
+
+        return coords1 - coords0, flow_up
```

Comment:
again, sorry about that but to a newcomer to this, naming makes this hard to follow :grimacing: 
like `network` could be `network_state`, it is not the actual network here. `input` could be `input_features`.
All across the code, `iters` should be something like `optical_flow_iterations` and should be defined in config. Here it has a default of 12, but the value of 20 is passed when `_forward` is actually called, the latter correct value should be in the config, makes everything cleaner!